Zend/zend_compile.c

```diff
@@ -6737,12 +6737,70 @@ zend_op *zend_compile_class_decl(zend_ast *ast, zend_bool toplevel) /* {{{ */
 			zend_error_noreturn(E_COMPILE_ERROR, "Clone method %s::%s() cannot be static",
 				ZSTR_VAL(ce->name), ZSTR_VAL(ce->clone->common.function_name));
 		} else if (ce->clone->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+			if (ZEND_TYPE_PURE_MASK(ce->clone->common.arg_info[-1].type) != MAY_BE_VOID) {
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Method %s::%s() must return void",
+					ZSTR_VAL(ce->name), ZSTR_VAL(ce->clone->common.function_name));
+			}
+		}
+	}
+
+	if (ce->__set && ce->__set->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		if (ZEND_TYPE_PURE_MASK(ce->__set->common.arg_info[-1].type) != MAY_BE_VOID) {
 			zend_error_noreturn(E_COMPILE_ERROR,
-				"Clone method %s::%s() cannot declare a return type",
-				ZSTR_VAL(ce->name), ZSTR_VAL(ce->clone->common.function_name));
+				"Setter %s::%s() must return void",
+				ZSTR_VAL(ce->name), ZSTR_VAL(ce->__set->common.function_name));
 		}
 	}
 
+	if (ce->__unset && ce->__unset->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		if (ZEND_TYPE_PURE_MASK(ce->__unset->common.arg_info[-1].type) != MAY_BE_VOID) {
+			zend_error_noreturn(E_COMPILE_ERROR,
+				"Method %s::%s() must return void",
+				ZSTR_VAL(ce->name), ZSTR_VAL(ce->__unset->common.function_name));
+		}
+	}
+
+	if (ce->__isset && ce->__isset->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		 if (ZEND_TYPE_PURE_MASK(ce->__isset->common.arg_info[-1].type) != MAY_BE_BOOL) {
```

Comment:
Looks like `ZEND_TYPE_PURE_MASK` does not check if the return type is **only** `bool`.

If I change the test `Zend/tests/return_types/034.phpt` to:

```diff
--TEST--
__isset can only declare a boolean return type
--FILE--
<?php

class Foo {
-	function __isset($name) : \stdClass {}
+	function __isset($name) : \stdClass|bool {}
}
?>
--EXPECTF--
Fatal error: Method Foo::__isset() must return a boolean value in %s on line %d
```

The test fails.